pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/KafkaAbstractSource.java

```diff
@@ -116,19 +117,21 @@ public void close() throws InterruptedException {
         LOG.info("Kafka source stopped.");
     }
 
+    @SuppressWarnings("unchecked")
     public void start() {
         runnerThread = new Thread(() -> {
-            LOG.info("Starting kafka source");
+            LOG.info("Starting kafka source on {}", kafkaSourceConfig.getTopic());
             consumer.subscribe(Collections.singletonList(kafkaSourceConfig.getTopic()));
             LOG.info("Kafka source started.");
-            ConsumerRecords<String, byte[]> consumerRecords;
             while (running) {
-                consumerRecords = consumer.poll(1000);
+                ConsumerRecords<Object, Object> consumerRecords = consumer.poll(1000);
                 CompletableFuture<?>[] futures = new CompletableFuture<?>[consumerRecords.count()];
                 int index = 0;
-                for (ConsumerRecord<String, byte[]> consumerRecord : consumerRecords) {
-                    LOG.debug("Record received from kafka, key: {}. value: {}", consumerRecord.key(), consumerRecord.value());
-                    KafkaRecord<V> record = new KafkaRecord<>(consumerRecord, extractValue(consumerRecord));
+                for (ConsumerRecord<Object, Object> consumerRecord : consumerRecords) {
+                    KafkaRecord record = new KafkaRecord(consumerRecord,
+                            extractKey(consumerRecord),
```

Comment:
any reason why do you extract the key right now? Why can't you extract the key lazily when it was accessed via KafkaRecord.getKey?